examples/gguf-hash/gguf-hash.cpp

```diff
@@ -0,0 +1,337 @@
+
+        char hex_result[41];
+        for (int  offset = 0; offset < 20; offset++) {
+            sprintf( ( hex_result + (2*offset)), "%02x", result[offset]&0xff);
+        }
+
+        printf("sha1    %s  %s\n", hex_result, fname.c_str());
+    }
+
+    if (hash_params.sha256) {
+        unsigned char result[SHA256_DIGEST_SIZE]; // sha256 outputs 32 bytes
+        sha256_final( &sha256_model_hash_ctx,  result);
+
+        char hex_result[SHA256_DIGEST_SIZE * 2 + 1] = {0};
+        for (int  offset = 0; offset < SHA256_DIGEST_SIZE; offset++) {
+            sprintf( ( hex_result + (2*offset)), "%02x", result[offset]&0xff);
+        }
+
+        printf("sha256  %s  %s\n", hex_result, fname.c_str());
+    }
+
+    ggml_free(ctx_data);
+    gguf_free(ctx);
+
+    return true;
+}
+
+static void generate_uuidv5(const unsigned char sha1_digest[20], unsigned char uuid[16]) {
+    // Ref: https://www.rfc-editor.org/rfc/rfc9562.html#section-5.5
+    // Assumes that digest was processed correctly with the expected namespace
+    for (int i = 0; i < 16; i++) {
+        uuid[i] = sha1_digest[i];
+    }
+
+    // Set bits corresponding to UUID ver 5
+    uuid[ 6] &= ~(0xF << 4);
+    uuid[ 6] |= (5 << 4);
+
+    // Set bits corresponding to UUID variant 0b10XX
+    uuid[ 8] &= ~(0xc << 4);
+    uuid[ 8] |= (0x8 << 4);
+}
+
+static bool gguf_uuid(const hash_params & hash_params) {
+    if (!hash_params.uuid) {
+        return true;
+    }
+
+    const std::string & fname = hash_params.input;
+    struct ggml_context * ctx_data = NULL;
+
+    struct gguf_init_params params = {
+        /*.no_alloc = */ false,
+        /*.ctx      = */ &ctx_data,
+    };
+
+    // sha1 init
+    SHA1_CTX sha1_model_hash_ctx;
+    SHA1Init(&sha1_model_hash_ctx);
+
+    unsigned char const uuidv5_namespace[] = {UUID_NAMESPACE_LLAMA_CPP_HEX};
+    SHA1Update( &sha1_model_hash_ctx, (unsigned char const *)uuidv5_namespace, sizeof(uuidv5_namespace));
+
+    struct gguf_context * ctx = gguf_init_from_file(fname.c_str(), params);
+    const int n_tensors = gguf_get_n_tensors(ctx);
+    for (int i = 0; i < n_tensors; ++i) {
+        const char * name = gguf_get_tensor_name(ctx, i);
+        struct ggml_tensor * cur = ggml_get_tensor(ctx_data, name);
+        auto n_bytes = ggml_nbytes(cur);
+        auto *raw_data = cur->data;
+        SHA1Update( &sha1_model_hash_ctx, (unsigned char const *)raw_data, n_bytes);
+    }
+
+    unsigned char result[21];
+    SHA1Final(result, &sha1_model_hash_ctx);
+
+    unsigned char uuid[16];
+    generate_uuidv5(result, uuid);
+
+    char string_buffer[37] = {0};
+    sprintf(string_buffer, "%02x%02x%02x%02x-%02x%02x-%02x%02x-%02x%02x-%02x%02x%02x%02x%02x%02x",
+        uuid[0], uuid[1], uuid[2], uuid[3],
+        uuid[4], uuid[5], uuid[6], uuid[7],
+        uuid[8], uuid[9], uuid[10], uuid[11],
+        uuid[12], uuid[13], uuid[14], uuid[15]);
+    printf("UUIDv5  %s  %s\n", string_buffer, fname.c_str());
```

Comment:
Does the UUID bring any value compared to the global hash? It's still a unique number, just a different format. It seems unnecessary

Reply:
Fyi just added validation capability.

As for the usefulness of uuid. In this context it's certainly not very useful as a hash🎇, instead it's mean to provide an option for models that do not have exisiting general.uuid in the kv store that I am proposing to add in a different PR... UUID is widely supported as a database key and I think it's helpful to be able to reference a Digital object via a unique ID.

In this case I am choosing to use UUID. Why not sha256 or sha1? Because it would either be too long and not as useful and as widely used. Plus I am hoping to be able to track gguf model parents as well and those will not always be  generatable ID but they can at least choose their own via random UUIDv4 generator.